src/org/traccar/protocol/Tk103ProtocolEncoder.java

```diff
@@ -16,36 +16,89 @@
  */
 package org.traccar.protocol;
 
+import org.traccar.Context;
 import org.traccar.StringProtocolEncoder;
 import org.traccar.helper.Log;
 import org.traccar.model.Command;
 
 public class Tk103ProtocolEncoder extends StringProtocolEncoder {
 
+    protected String t580WFormatCommand(Command command, String format, String... keys) {
+        String content = super.formatCommand(command, format, keys);
+        return String.format("[begin]sms2,%s,[end]", content);
+    }
+
+    protected String checkEn(Command command, String ifEnabled, String ifDisabled) {
+        return command.getBoolean(Command.KEY_ENABLE) ? ifEnabled : ifDisabled;
+    }
+
     @Override
     protected Object encodeCommand(Command command) {
 
-        switch (command.getType()) {
-            case Command.TYPE_GET_VERSION:
-                return formatCommand(command, "({%s}AP07)", Command.KEY_UNIQUE_ID);
-            case Command.TYPE_REBOOT_DEVICE:
-                return formatCommand(command, "({%s}AT00)", Command.KEY_UNIQUE_ID);
-            case Command.TYPE_SET_ODOMETER:
-                return formatCommand(command, "({%s}AX01)", Command.KEY_UNIQUE_ID);
-            case Command.TYPE_POSITION_SINGLE:
-                return formatCommand(command, "({%s}AP00)", Command.KEY_UNIQUE_ID);
-            case Command.TYPE_POSITION_PERIODIC:
-                return formatCommand(command, "({%s}AR00%s0000)", Command.KEY_UNIQUE_ID,
-                        String.format("%04X", command.getInteger(Command.KEY_FREQUENCY)));
-            case Command.TYPE_POSITION_STOP:
-                return formatCommand(command, "({%s}AR0000000000)", Command.KEY_UNIQUE_ID);
-            case Command.TYPE_ENGINE_STOP:
-                return formatCommand(command, "({%s}AV011)", Command.KEY_UNIQUE_ID);
-            case Command.TYPE_ENGINE_RESUME:
-                return formatCommand(command, "({%s}AV010)", Command.KEY_UNIQUE_ID);
-            default:
-                Log.warning(new UnsupportedOperationException(command.getType()));
-                break;
+        boolean deviceT580W = Context.getIdentityManager().lookupAttributeBoolean(
```

Comment:
I would recommend to use something more generic instead of model name. I think `alternative` as we use in GT06 would be better and more consistent.

Reply:
Thanks for the tip! I will implement exactly the same approach as it was done in GT06.